### ISSUE TITLE
Expected performance in a real-time plot

### ISSUE THREAD

**Issue by raffaeler** [oxyplot/oxyplot#1801]
I am acquiring 50 samples/sec from a board and willing to display 1h data points window, by trimming the data as soon as the acquisition goes beyond the given window. The data is made of milliseconds and voltage (XY).

The tests I made are the following:
1. For each point, putting a sample in a list, trimming the oldest (if any) and Invalidate the plot
2. Same as above but binding an `ObservableCollection` to the `ItemsSource`
3. Creating a `TimeSpanAxis` and varying the Minimum and Maximum values in binding in order to "view" a portion of the data stored in a simple list.

As soon as the samples go beyond 5000, the application is almost impossible to use (too slow). In the best case (number 1) the app has hiccups every 5-10 seconds.
What is the suggested strategy to get a decent performance?
Is this a supported scenario or should I use a different library?

TIA

**Comment by alxkalinin** [comment]
I have a quite good performance in WPF real-time plotting with FPS=60 up to 100000 point in series. I use [RealtimeDemo](https://github.com/oxyplot/oxyplot/tree/develop/Source/Examples/WPF/WpfExamples/Examples/RealtimeDemo) example as a base. 

My strategy for real-time rendering is
- I have my own data array and I update data array not in UI thread
- OxyPlot series contains fixed number of points, not growing list
- I use 'CompositionTarget.Rendering" UI thread event for updating a plot; in this event I select a data range for displaying and copy this data range from my data array to the fixed size OxyPlot series

**Comment by Sandr0oo** [comment]
@alxkalinin Can you give me a simple example? Я не понимаю про "CompositionTarget.Rendering" и "copy this data range from my data array to the...".

**Comment by Sandr0oo** [comment]
@alxkalinin What if I need non-static axes? I've noticed that lags on the graphics appear if I resize my application to fullscreen. The smaller the window, the less lag.

**Comment by alxkalinin** [comment]
@Sandr0oo 

> I've noticed that lags on the graphics appear if I resize my application to fullscreen. The smaller the window, the less lag.

Do you have a GPU on your machine?

In the WPF we have two threads:
- **UI thread**, running on CPU
- **Rendering thread**, running on GPU with the help of Direct3D

The UI thread make a _composition_, it means that the UI thread put all vector graphical object into the scene. The Rendering thread takes the composition of the vector objects and produces the actual visual screen.

Large windows should increase GPU times, not CPU. And I don't worry to much about GPU time. On a typical machine GPU is not a bottleneck.

The UI thread performance does not depend on the screen resolution. Only Rendering thread can be slow on large windows.

> What if I need non-static axes?

In that case you also need to update an axes plot. But here you have a flexibility. You have independently
- your plot
- your axes
- composition rendering event, with the FPS 62, when a composition is ready to be send to a rendering and you can add your own object to the composition

You can, for example, update your axes not every frame, but each 2 or 3 frames. Or you can reduce number of point to be added to the plot on each frame.

With all these techniques the real-time rendering is quite fast and do not eat too many time from the UI thread. That is my experience.

**Comment by rs38** [comment]
using WPF I am not able to automatically update the plot by assigning an ObservableCollection to the ItemsSource of LineSeries without using `InvalidatePlot`.
what am I missing? Has some behavious changed?

**Comment by VisualMelon** [comment]
@rs38 as far as I'm aware, that has never been supported (though I've not been around since much before 2.0)

**Comment by rs38** [comment]
it's quite hard for a WPF novice to keep track as the docs https://oxyplot.readthedocs.io/en/latest/ are very outdated and most google findings as well?! like mentioning XAML axis definitions 
```
<oxy:Plot Title="{Binding Title}" InvalidateFlag="{Binding InvalidateFlag}">
    <oxy:Plot.Series>
...
```
and SO as well:
https://stackoverflow.com/questions/31851797/oxyplot-adding-data-from-an-array
https://stackoverflow.com/questions/50649467/oxyplot-graphic-doesnt-update

**Comment by VisualMelon** [comment]
@rs38 yes, apologies for the confusion. The XAML provision does I believe support it; however, it has moved to https://github.com/oxyplot/oxyplot-contrib due to maintenance concerns.

If you believe it is not working, it's best to open a new issue over there.

The documentation certainly does need updating.
